### PR TITLE
feat(new source): Initial `splunk_hec` source implementation

### LINKED ISSUE #1088: New `splunk_hec` source
### Description

A new source that partially implements Splunk HTTP Event Collector REST API. This `splunk_hec source` will be able to recieve logs in splunk format.

### Motivation

One of use cases for this is being able to collect logs from Docker over `splunk` log driver. This is usefull in constrained enviorments like ECS where Fargate only supports three log drivers, one AWS propriatery, one custom for fluentd, and splunk. (#813)

### Prior art

Vector already has [`splunk_hec sink`](https://docs.vector.dev/usage/configuration/sinks/splunk_hec). Implementation of this will be it's counterpart.

### Behavior

In short, `splunk_hec source` will listen for incoming `http` POST request, and construct log events from them.

Out of whole [Splunk HEC REST API](https://docs.splunk.com/Documentation/Splunk/7.3.2/Data/HECRESTendpoints), access points to be supported will be:
 - services/collector/event/1.0, for receiving logs in `inline` and `json` format
 - services/collector/raw/1.0, for receiving logs in `raw` format
 - services/collector/health, for passing health checks

Additionally:
 - `gzip`
 - `http` token authorization

will be supported.

This is a minimal set of functionalities that enable `splunk_hec source` to fully support receiving logs from Docker `splunk` log driver, Vector` splunk_hec sink`, and any other source which uses HEC only as a destination for logs.

### Configuration

```
[sources.my_source_id]
  type = "splunk_hec" 

  # The address to bind the socket to.
  # 
  # * optional
  # * default 0.0.0.0:8088
  address = "0.0.0.0:8088"

  # Splunk HEC token.
  # 
  # * required
  # * no default
  token = "A94A8FE5CCB19BA61C4C08"
```

### Implementation

Access point services/collector/event/1.0, will parse body of the request with `json` parser whose results will be flattened and turned into log events.

Access point services/collector/raw/1.0, will create a log event with body of the request as a `message`.

### PR DESCRIPTION
**Pull request by ktff**
Closes #1088 

### Changes

`address` field in configuration has a default value of `0.0.0.0:8088` since port 8088 is a default port for Splunk HEC servers.

### Testing

In code, tests are using `splunk_hec` sink to test the source. But besides that, this source should be tested with other clients/environments:

- [x]  Docker
- ECS
     - [x] Fargate
     - [x] EC2

### CODE REVIEW

**Comment by binarylogic** [comment]
Thanks @ktff, this looks great. I'm assigning @Jeffail for review only. We should have this reviewed by next week (hopefully this week).

File: src/sources/splunk_hec.rs [hunk 1]
```diff
@@ -0,0 +1,954 @@
...
+    // Add host
+    if let Some(host) = host {
+        log.insert_explicit(event::HOST.clone(), host.as_bytes().into());
+    }
+
+    Ok(event)
+}
+
+/// Recursevly inserts json values to event under given name
+pub fn insert(event: &mut LogEvent, name: String, value: Value) {
```
Comment (Jeffail) [inline src/sources/splunk_hec.rs]: This seems generally useful, might be worth placing this within the event module as a method. It'd also make it easier to find when https://github.com/timberio/vector/issues/704 gets worked on.

File: src/sources/splunk_hec.rs [hunk 2]
```diff
@@ -0,0 +1,954 @@
...
+
+// Event fields unique to splunk_hec source
+lazy_static! {
+    pub static ref CHANNEL: Atom = Atom::from("channel");
+    pub static ref INDEX: Atom = Atom::from("index");
+    pub static ref SOURCE: Atom = Atom::from("source");
+    pub static ref SOURCETYPE: Atom = Atom::from("sourcetype");
+}
+
+/// Cashed bodies for common responses
```
Comment (LucioFranco) [inline src/sources/splunk_hec.rs]:
```suggestion
/// Cached bodies for common responses
```

is this what you meant?

File: src/sources/splunk_hec.rs [hunk 3]
```diff
@@ -0,0 +1,954 @@
...
+
+        let source = Arc::new(SplunkSource::new(self, out, trigger));
+
+        let service = move || future::ok::<_, String>(Connection::new(source.clone()));
+
+        // Build server
+        let server = Server::bind(&self.address)
+            .serve(service)
+            .with_graceful_shutdown(tripwire)
+            .map_err(|error| error!(message="Splunk HEC source stopped, because of error", %error));
```
Comment (LucioFranco) [inline src/sources/splunk_hec.rs]:
```suggestion
            .map_err(|error| error!(message="Splunk HEC source stopped.", %error));
```

the `error!` macro should be enough to inform users it was an error :)

**Comment by LucioFranco** [comment]
At a high level this seems fine, but I am curious if there was a reason we were not able to use something like warp to power the webserver side?

File: src/sources/splunk_hec.rs (hunk 2 above)
Comment (ktff) [reply src/sources/splunk_hec.rs]: Yep

File: src/sources/splunk_hec.rs (hunk 3 above)
Comment (ktff) [reply src/sources/splunk_hec.rs]: True :D

**Comment by ktff** [comment]
I honestly forgot to look beyond hyper as it fits the purpose quite nicely. 

Some 100-150 lines of code could be abstracted to some higher level library. 

For warp, there would be some ~100 lines less.

@LucioFranco do you think it would be worth it?

File: src/sources/splunk_hec.rs (hunk 1 above)
Comment (ktff) [reply src/sources/splunk_hec.rs]: I agree. 

Also this is actually a modified version of basically the same function in `json_parser`, so I will also adjust `json_parser` to use that.

File: src/sources/splunk_hec.rs [hunk 4]
```diff
@@ -0,0 +1,888 @@
...
+    /// Remaining request with JSON events
+    data: R,
+    /// Count of sended events
+    events: usize,
+    /// Optinal channel from headers
+    channel: Option<ValueKind>,
+    /// Extracted default time
+    time: Option<ValueKind>,
+    /// Remaining extracted default values
+    extractors: [DefaultExtractor; 4],
```
Comment (LucioFranco) [inline src/sources/splunk_hec.rs]: I might be missing something but can you explain what these used for?

**Comment by LucioFranco** [comment]
@a-rodin looks like CI got sigkilled any idea why?

File: src/sources/splunk_hec.rs (hunk 4 above)
Comment (ktff) [reply src/sources/splunk_hec.rs]: Spluk HEC `api` specifies for some `json` fields that once they are seen their value becomes default for next events in the request until that field is found again. `DefaultExtractor` implements that behavior.

**Comment by Jeffail** [comment]
Merged (with collisions fixed): https://github.com/timberio/vector/commit/a68c9781a12cd35f2ee1cd7686320d1bd6e52c05

**Pull request closed without merging**